### PR TITLE
EMS coil speed actuator not working for multispeed coils in UnitarySystem

### LINKED ISSUE #9732: Using actuator Unitary System DX Coil Speed Value doesn't report correctly in the latest release
Issue overview
--------------
I surprisingly found that whenever an EMS program using following speed level actuator will remove the output variables: ``Output:Variable,*,Cooling Coil Crankcase Heater Electricity Rate,hourly; !- HVAC Average [W]`` and ``Output:Variable,*,Cooling Coil Crankcase Heater Electricity Energy,hourly; !- HVAC Sum [J]`` from idd file, and a warning of requesting unexpected output variables will be thrown.

EnergyManagementSystem:Actuator,
  central_ac_unitary_system_speed_override, !- Name
  central ac unitary system,              !- Actuated Component Unique Name
  Coil Speed Control,                     !- Actuated Component Type
  Unitary System DX Coil Speed Value;     !- Actuated Component Control Type

Attached are two idfs that the only difference is a dumb EMS program and the RDD files generated.
[in-noEMS.txt](https://github.com/NREL/EnergyPlus/files/9984983/in-noEMS.txt)
[in-dumbSpeedLevelActuator.txt](https://github.com/NREL/EnergyPlus/files/9984984/in-dumbSpeedLevelActuator.txt)
[in-dumbSpeedLevelActuator-rdd.txt](https://github.com/NREL/EnergyPlus/files/9984992/in-dumbSpeedLevelActuator-rdd.txt)
[in-noEMS-rdd.txt](https://github.com/NREL/EnergyPlus/files/9984993/in-noEMS-rdd.txt)

More descriptions here: [Link](https://github.com/NREL/OpenStudio-HPXML/pull/744#issuecomment-1309665978)

### Details
Some additional details for this issue (if relevant):
 - Platform (Windows, 11)
 - Version of EnergyPlus (22.2.0)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #9732
 - The coil model was assuming there were a cooling and heating coil present during certain calculations. Additionally, the coil model was not called when EMS was used to actuate the multisped cooling coil.

Various items corrected:

- The crankcase report variables were not enabled for Coil:Cooling:DX:MultiSpeed.
- The UnitarySystem was not calling the coil model when EMS speed control was used.
- The air flow rate used for multi-speed coils was not correct for cycling fan mode, the upper speed flow rate was used instead of a PLR average of upper and lower speed air flow rate.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by rraustad** [comment]
There were several things going on here:
1) The crankcase report variables were not enabled for Coil:Cooling:DX:MultiSpeed.
2) Some calculations assumed the cooling and heating coils were present and caused a hard crash.
3) The UnitarySystem does not operate when there is no cooling or heating load, regardless of EMS controls. This may or may not be the intended operation but that control will not be changed here.
4) The UnitarySystem was not calling the coil model when EMS speed control was used. This change seems strange since there is a unit test that seems to prove otherwise, although  in that unit test the cooling coil may not have been active (I didn't check that).
5) The defect file does not have a cooling load and therefore the cooling coil is inactive. The modified defect file does show cooling coil operation.
6) The defect file crankcase reports are now available in rdd file.

Note the hidden rows in this spreadsheet to show multiple operational controls. Crankcase heater operates in winter. No crankcase heater in summer when OAT is higher than maximum OAT for crankcase heater operation. The cooling coil operates as specified by EMS speed control, only when a cooling load exists.

image

Modified defect file used for the above output:

[in-dumbSpeedLevelActuator_modified.idf.txt](https://github.com/NREL/EnergyPlus/files/10156395/in-dumbSpeedLevelActuator_modified.idf.txt)

**Comment by rraustad** [comment]
The diffs in several files will need to be investigated. I would think there would be no diffs if EMS speed control was not used.

File: src/EnergyPlus/DXCoils.cc [hunk 1]
```diff
@@ -6489,6 +6489,28 @@ void GetDXCoils(EnergyPlusData &state)
...
+                                    OutputProcessor::SOVTimeStepType::System,
+                                    OutputProcessor::SOVStoreType::Summed,
+                                    Coil.Name,
+                                    _,
+                                    "Electricity",
+                                    "DHW",
+                                    _,
+                                    "Plant");
+            }
+
```
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc]: Enable crankcase reports for Coil:Cooling:DX:MultiSpeed

File: src/EnergyPlus/UnitarySystem.cc [hunk 2]
```diff
@@ -2094,19 +2094,23 @@ namespace UnitarySystems {
...
-                    NoLoadHeatingAirFlowRateRatio =
-                        min(this->m_NoLoadAirFlowRateRatio, state.dataUnitarySystems->designSpecMSHP[MSHPIndex].heatingVolFlowRatio[0]);
+                if (state.dataUnitarySystems->designSpecMSHP[MSHPIndex].numOfSpeedHeating > 0) {
+                    if (state.dataUnitarySystems->designSpecMSHP[MSHPIndex].heatingVolFlowRatio[0] == DataSizing::AutoSize) {
+                        NoLoadHeatingAirFlowRateRatio =
+                            min(this->m_NoLoadAirFlowRateRatio, 1.0 / double(state.dataUnitarySystems->designSpecMSHP[MSHPIndex].numOfSpeedHeating));
+                    } else {
+                        NoLoadHeatingAirFlowRateRatio =
+                            min(this->m_NoLoadAirFlowRateRatio, state.dataUnitarySystems->designSpecMSHP[MSHPIndex].heatingVolFlowRatio[0]);
+                    }
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: More protection for when coil are not present via the multispeed performance object input for num speeds

File: src/EnergyPlus/UnitarySystem.cc [hunk 3]
```diff
@@ -8392,8 +8396,12 @@ namespace UnitarySystems {
...
         Real64 CoolingOnlySensibleOutput =
             state.dataLoopNodes->Node(this->CoolCoilInletNodeNum).MassFlowRate * CpAir *
-            ((state.dataLoopNodes->Node(this->NodeNumOfControlledZone).Temp - state.dataLoopNodes->Node(this->CoolCoilOutletNodeNum).Temp) -
-             (state.dataLoopNodes->Node(this->HeatCoilOutletNodeNum).Temp - state.dataLoopNodes->Node(this->HeatCoilInletNodeNum).Temp));
+            (state.dataLoopNodes->Node(this->NodeNumOfControlledZone).Temp - state.dataLoopNodes->Node(this->CoolCoilOutletNodeNum).Temp);
+        if (this->m_HeatCoilExists) {
+            CoolingOnlySensibleOutput -=
+                state.dataLoopNodes->Node(this->CoolCoilInletNodeNum).MassFlowRate * CpAir *
+                (state.dataLoopNodes->Node(this->HeatCoilOutletNodeNum).Temp - state.dataLoopNodes->Node(this->HeatCoilInletNodeNum).Temp);
+        }
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Calculation assumed coils were present, added additional protection.

File: src/EnergyPlus/UnitarySystem.cc [hunk 4]
```diff
@@ -11552,29 +11560,32 @@ namespace UnitarySystems {
         } else {
             OutsideDryBulbTemp = state.dataEnvrn->OutDryBulbTemp;
         }
-        if (this->m_EMSOverrideCoilSpeedNumOn) {
-            // pass
-        } else {
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: This is a big change. The coil model was not called when EMS coil speed actuator was used.

File: src/EnergyPlus/UnitarySystem.cc [hunk 5]
```diff
@@ -11591,195 +11602,194 @@ namespace UnitarySystems {
...
+
+            this->m_CoolCompPartLoadRatio = (CompressorOn == DataHVACGlobals::CompressorOperation::On) ? PartLoadRatio : 0.0;
+        } break;
+        case DataHVACGlobals::CoilDX_PackagedThermalStorageCooling: {
+            PackagedThermalStorageCoil::SimTESCoil(state, CompName, this->m_CoolingCoilIndex, this->m_FanOpMode, this->m_TESOpMode, PartLoadRatio);
+        } break;
+        default:
+            break;
         }
+        //        }
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:11792]: End of IF block when EMS coil speed control actuator is used.

File: src/EnergyPlus/UnitarySystem.cc [hunk 6]
```diff
@@ -12615,6 +12625,9 @@ namespace UnitarySystems {
                     } else {
                         doIt = true;
                     }
+                } else if (this->m_EMSOverrideCoilSpeedNumOn &&
+                           (CoilType_Num == DataHVACGlobals::CoilDX_MultiSpeedCooling || CoilType_Num == DataHVACGlobals::CoilDX_Cooling)) {
+                    // do nothing, PartLoadFrac set above
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Load based controls were overwriting EMS speed control actuator

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc [hunk 7]
```diff
@@ -17398,6 +17406,27 @@ TEST_F(ZoneUnitarySysTest, UnitarySystemModel_MultiSpeedDXCoilsDirectSolutionTes
...
+                      OAUCoilOutTemp,
+                      ZoneEquipment,
+                      sensOut,
+                      latOut);
+    EXPECT_NEAR(thisSys->m_CycRatio, 1.00, 0.01);
+    EXPECT_NEAR(thisSys->m_SpeedRatio, 0.087, 0.01);
+    EXPECT_NEAR(thisSys->m_CoolingSpeedRatio, 0.087, 0.01);
+    EXPECT_EQ(thisSys->m_CoolingSpeedNum, 2);
+    EXPECT_NEAR(sensOut, -11998.0, 3.0);
+
```
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc]: Added missing unit test configuration.

    // thisSys->m_EMSOverrideCoilSpeedNumOn = true, state->dataGlobal->DoCoilDirectSolutions = true

File: src/EnergyPlus/DXCoils.cc [hunk 8]
```diff
@@ -6489,6 +6489,28 @@ void GetDXCoils(EnergyPlusData &state)
...
+                                    Coil.Name);
+                SetupOutputVariable(state,
+                                    "Cooling Coil Crankcase Heater Electricity Energy",
+                                    OutputProcessor::Unit::J,
+                                    Coil.CrankcaseHeaterConsumption,
+                                    OutputProcessor::SOVTimeStepType::System,
+                                    OutputProcessor::SOVStoreType::Summed,
+                                    Coil.Name,
+                                    _,
+                                    "Electricity",
```
Comment (amirroth) [inline src/EnergyPlus/DXCoils.cc]: 📌 "Electricity", "DHW", are "Plant" are all enumerated values (ResourceType/FuelType, EndUse, MeterSuperGroup), correct? You know what I am going to say then.

Comment (rraustad) [reply src/EnergyPlus/DXCoils.cc]: Never do today what you can do tomorrow. Something may occur to make you regret your premature action. Aaron Burr.

Comment (amirroth) [reply src/EnergyPlus/DXCoils.cc:6508]: `(lambda f.(lambda x.f(x x) (lambda x.f(x x))) lambda y.("Never do y what you can do y+1"))` --Haskell Curry.

Comment (amirroth) [reply src/EnergyPlus/DXCoils.cc]:
```cpp
while (true) {
   if (can_do_x_tomorrow) {
       sleep(24*60*60);
       continue;
   } else {
       do_x();
       break;
   }
}
```

**Comment by rraustad** [comment]
The failed unit test and at least the example file EMSSetpointBasedMultiSpeedDXCoilOverrideControl were not setting the correct air flow rate for cycling fan mode. A comparison of this example file shows the zone temp to be now at set point. More diffs review is needed.

image

File: src/EnergyPlus/UnitarySystem.cc [hunk 9]
```diff
@@ -10952,8 +10960,14 @@ namespace UnitarySystems {
...
-                AverageUnitMassFlow = FanPartLoadRatio * state.dataUnitarySystems->CompOnMassFlow +
-                                      (1.0 - FanPartLoadRatio) * state.dataUnitarySystems->CompOffMassFlow;
+                if (this->m_FanOpMode == DataHVACGlobals::ContFanCycCoil) {
+                    AverageUnitMassFlow = FanPartLoadRatio * state.dataUnitarySystems->CompOnMassFlow +
+                                          (1.0 - FanPartLoadRatio) * state.dataUnitarySystems->CompOffMassFlow;
+                } else {
+                    Real64 tempSpeedRatio = state.dataUnitarySystems->CoolingLoad ? this->m_CoolingSpeedRatio : this->m_HeatingSpeedRatio;
+                    AverageUnitMassFlow = tempSpeedRatio * state.dataUnitarySystems->CompOnMassFlow +
+                                          (1.0 - tempSpeedRatio) * state.dataUnitarySystems->CompOffMassFlow;
+                }
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: setAverageAirFlow was not handling the average air flow calculation correctly for multipeed coils for cycling fan mode. I wonder if there are certain other configurations where this math is incorrect?  e.g., single mode switch

**Comment by nrel-bot-3** [comment]
@rraustad @myoldmopar it has been 28 days since this pull request was last updated.

**Comment by yzhou601** [comment]
@rraustad Hi, Richard, it seems it has been a while since your last commit, is this PR good for test? We do have some test files that we want to run to make sure it's working correctly.

**Comment by rraustad** [comment]
@yzhou601 I believe these were the only changes needed. The diffs need to be reviewed again since I can't recall what the diffs showed. Yes, please run your test files against this branch to provide additional testing.

**Comment by rraustad** [comment]
One of the problems I keep running into is function setOnOffMassFlowRate doesn't account for constant fan in some places, maybe this is the last place needing correction, maybe not. I think if we remove MaxCoolMassFlowRate (and MaxHeat) and the corresponding volume flow rate variables, and instead use the m_CoolMassFlowRate[1] for all coil types (single speed coils would have MaxSpeeds = 1), the setOnOffMassFlowRate function would be much cleaner. That change is for another day since it may also affect the logic flow in the control routines.

**Comment by Myoldmopar** [comment]
@yzhou601 when you have a chance to run your test files, please report back and we'll work to get this merged if you feel like it is solving the issue for you.

**Comment by yzhou601** [comment]
Thanks @rraustad for the bugfix!! I've run several test files on my end, E+ built from this branch no longer complains about crankcase energy output variable. And the staging is working again, this looks pretty good to me. 
See below one of the tested time-based staging control and its energy results. 

image
image

A quick question, maybe not related to this PR: 

I noticed that the `Unitary System Sensible Cooling Rate` is much lower than `Cooling Coil Sensible Cooling Rate`, and the difference is larger than fan heat gain to air, any thought what can be the additional heat source in the system?

**Comment by rraustad** [comment]
> Unitary System Sensible Cooling Rate is much lower than Cooling Coil Sensible Cooling Rate:

I would guess the difference is the outdoor air entering the unit. Both calculations should be Tout - Tin (instead of Tzone), so the coil sensible capacity should equal system sensible capacity + fan heat + outdoor air sensible load. If not then we should look into that discrepancy. Since Tout is the same, look at Tcoil,in versus Tunitarysystem, in (and win used for Cp calculation). The difference should be clear (i.e., ratio of delta Ts should nearly equal ratio of sensible cooling rates, with an accounting of w used for Cp).

    void UnitarySys::calculateCapacity(EnergyPlusData &state, Real64 &SensOutput, Real64 &LatOutput)
    {

        // Check delta T (outlet to reference temp), IF positive use reference HumRat ELSE outlet humrat to calculate
        // sensible capacity as MdotDeltaH at constant humidity ratio
        int OutletNode = this->AirOutNode;
        Real64 AirMassFlow = state.dataLoopNodes->Node(OutletNode).MassFlowRate;
        Real64 RefTemp = 0.0;
        Real64 RefHumRat = 0.0;
        if (this->m_ControlType == UnitarySysCtrlType::Setpoint) {
            RefTemp = state.dataLoopNodes->Node(this->AirInNode).Temp;           <------------
            RefHumRat = state.dataLoopNodes->Node(this->AirInNode).HumRat;
        } else {
            RefTemp = state.dataLoopNodes->Node(this->NodeNumOfControlledZone).Temp;
            RefHumRat = state.dataLoopNodes->Node(this->NodeNumOfControlledZone).HumRat;
        }

**Comment by rraustad** [comment]
@yzhou601 thanks for the additional testing. For the plot, I would also be interested in Tout versus Tout, setpoint for set point control. If load control, then Tzone versus Tstat temp.

**Comment by yzhou601** [comment]
> @yzhou601 thanks for the additional testing. For the plot, I would also be interested in Tout versus Tout, setpoint for set point control. If load control, then Tzone versus Tstat temp.

Hi @rraustad , yes it is setpoint control. See below the requested temperature plot.
image
The system layout is Fan -> Cooling Coil. So the nodes are ordered as: UnitarySystemInlet(FanInlet) -> FanOutlet(CoolingCoilInlet) -> CoolingCoilOutlet(UnitarySystemOutlet), Tout should be the same node temperature, but for Tin, Coil is higher than unitary system, which makes sense because of fan, but the question is that fan power doesn't solely account for the sensible cooling energy difference between unitary system output and coil output. 
I don't think we have implemented any outdoor air equipment in unitary system, but there're `ZoneInfiltration:DesignFlowRate` objects which are actuated by EMS program, but I assume it's not related to the system energy calculation?
I have attached this test file if you want to quickly explore more.
[in.idf.txt](https://github.com/NREL/EnergyPlus/files/10582182/in.idf.txt)

**Comment by rraustad** [comment]
Oh yeah, the UnitarySystem doesn't have the OA mixer. I just checked and it looks llke the UnitarySystem uses Win and the coil uses Wout in the sensible calculation. That would explain the difference you see. This is a separate issue from this control problem.

Here is the call in UnitarySystem:

    CalcZoneSensibleLatentOutput(AirMassFlow,
                                 state.dataLoopNodes->Node(OutletNode).Temp,
                                 state.dataLoopNodes->Node(OutletNode).HumRat,
                                 RefTemp,
                                 RefHumRat,

And in this function the 5th argument is used, which for set point control, is the inlet node w. The coil model calls `CalcComponentSensibleLatentOutput` which uses the min of these w inputs.

    void CalcZoneSensibleLatentOutput(Real64 const MassFlow,  // air mass flow rate, {kg/s}
                                  Real64 const TDBEquip,  // dry-bulb temperature at equipment outlet {C}
                                  Real64 const WEquip,    // humidity ratio at equipment outlet
                                  Real64 const TDBZone,   // dry-bulb temperature at zone air node {C}
                                  Real64 const WZone,     // humidity ratio at zone air node
    )
    {
    if (MassFlow > 0.0) {
        TotalOutput = MassFlow * (Psychrometrics::PsyHFnTdbW(TDBEquip, WEquip) -
                                  Psychrometrics::PsyHFnTdbW(TDBZone, WZone));                         // total addition/removal rate, {W};
        SensibleOutput = MassFlow * Psychrometrics::PsyDeltaHSenFnTdb2Tdb1W(TDBEquip, TDBZone, WZone); // sensible addition/removal rate, {W};

**Comment by yzhou601** [comment]
> Oh yeah, the UnitarySystem doesn't have the OA mixer. I just checked and it looks llke the UnitarySystem uses Win and the coil uses Wout in the sensible calculation. That would explain the difference you see. This is a separate issue from this control problem.
> 
> Here is the call in UnitarySystem:
> 
> ```
> CalcZoneSensibleLatentOutput(AirMassFlow,
>                              state.dataLoopNodes->Node(OutletNode).Temp,
>                              state.dataLoopNodes->Node(OutletNode).HumRat,
>                              RefTemp,
>                              RefHumRat,
> ```
> 
> And in this function the 5th argument is used, which for set point control, is the inlet node w. The coil model calls `CalcComponentSensibleLatentOutput` which uses the min of these w inputs.
> 
> ```
> void CalcZoneSensibleLatentOutput(Real64 const MassFlow,  // air mass flow rate, {kg/s}
>                               Real64 const TDBEquip,  // dry-bulb temperature at equipment outlet {C}
>                               Real64 const WEquip,    // humidity ratio at equipment outlet
>                               Real64 const TDBZone,   // dry-bulb temperature at zone air node {C}
>                               Real64 const WZone,     // humidity ratio at zone air node
> )
> {
> if (MassFlow > 0.0) {
>     TotalOutput = MassFlow * (Psychrometrics::PsyHFnTdbW(TDBEquip, WEquip) -
>                               Psychrometrics::PsyHFnTdbW(TDBZone, WZone));                         // total addition/removal rate, {W};
>     SensibleOutput = MassFlow * Psychrometrics::PsyDeltaHSenFnTdb2Tdb1W(TDBEquip, TDBZone, WZone); // sensible addition/removal rate, {W};
> ```

So which approach is more correct? Should we use unitary system output or coil output for sensible cooling energy? Does the discrepancy also happen in heating? (I assume not, since there's no latent/sensible being calculated?)

**Comment by rraustad** [comment]
Let's see what #9833 shows wrt diffs.

**Comment by rraustad** [comment]
This test file is load control. In that case the reference is the zone condition and the unitary system sensible capacity represents the delivered capacity. This should be the same as the zone predicted load (if zone is at Tstat set point) and the ADU sensible capacity. This discussion makes me think the unitary system needs another sensible capacity report since this one switches reference between load and set point control.

1.31.2.2.9 Unitary System Sensible Cooling Rate [W]
This output field reports the moist air sensible heat extraction rate of the unitary system from the
zones it is serving in Watts. For set point control, this value is calculated using the enthalpy difference
of the unitary system outlet air and inlet air streams at a constant humidity ratio, and the air mass flow
rate through the unitary system. This value is calculated for each HVAC system timestep being simulated,
and the results (cooling only) are averaged for the timestep being reported. For load or single zone VAV
control this value is calculated using the outlet air and zone air conditions.

P.S. #9833 only changed the sensible capacity by 1 or 2 watts.

image

**Comment by Myoldmopar** [comment]
@rraustad with #9833 closed, do you have any updated comments for this PR?

**Comment by yzhou601** [comment]
@rraustad @Myoldmopar This was tested to be good on my end. Feel free to review, merge and make it into the upcoming release, thanks!

**Comment by Myoldmopar** [comment]
There's a small conflict here.  I assume it just occurred with your updates to UnitarySystem @rraustad.  Do you have a chance to look at the conflicts?  If not, or if you want me to assist, just let me know and I'll pull the branch and resolve them.

**Comment by Myoldmopar** [comment]
@rraustad it seems like the code here is pretty well stabilized.  I pulled the branch and was going to resolve the conflicts, but I'm not immediately clear which portions to keep, remove, etc.  Could you take a stab at it and let me know if there is anything left here before final review/merge?  If not, let me know and I'll give it another shot.  I may end up just "cherry picking" your meaningful changes if I go about it though.

Thanks!

**Comment by Myoldmopar** [comment]
Awesome, pulling and testing, thanks @rraustad !

**Comment by rraustad** [comment]
The conflicts were in the big chunk of code around 11600 where I removed:

        if (this->m_EMSOverrideCoilSpeedNumOn) {
            // pass
        } else {

I went line by line with develop and this looks the same as develop so I am confident in the conflict resolution.

**Comment by Myoldmopar** [comment]
Thanks @rraustad, I knew you would see exactly what to do with the conflict.  I've got it building and testing locally, I expect this to merge soon.

**Comment by Myoldmopar** [comment]
Yeah this looks good locally, no need to wait on further tests.  Thanks @rraustad (Decent CI may complain later since I'll delete the branch to save CI cycles, just ignore.)

**Pull request merged**